### DISCUSSION TITLE
ESP32cam with AM312

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#10733] [Using MicroPython]
Currently running some fairly old LeMaRamiva upython (micropython_camera_feeeb5ea3_esp32_idef4_4.bin) on one of those ubiquitous ESP32cam boards. An AM312 pyro sensor wakes it from deepsleep to take a pic. The AM312 works OK indoors but is prone to false triggering outdoors which got me thinking about some basic image processing (basic as in "how different is this pic from the last one?").

Smallish (~100k) packages like https://github.com/v923z/micropython-ulab look enticing but my chances of compiling them in with the LeMaRamiva port are not good.

How hard might it be to run the mean squared error equation (equation 1) @ https://pyimagesearch.com/2014/09/15/python-compare-two-images/ with the standard maths available in upython?  I had a few goes at it but my grasp of byte arrays is not good enough.

**Comment by 2dof** [comment]
MSE of 2 images you will find in   :https://datagy.io/mean-squared-error-python/ 
 ( not need to convert to float like in example). Note that MSE is sensitive to image brightens.
 
depends of Your goal to detect difference in your image, for example detect some move on image/object, the simplest way is:
image to greyscale ( if rgb), -> perform absolute difference ->  do normalization and some morphological operations -> use some threshold (area checking) to detect object ( usually we need only detect object in some section of our image so depends on using mask or best only our area of interest (only  part of image) we can speed up calculation, also if not need to preserve one of image, then calculation can be done 'in place'  to save memory space). 

edit. images are just 3D/2D arrays/matrix so ulab has all what You need.  if You can't use ulab  then In standard micro python where array.array()  can be only 1D  so converting image to 1D array is needed ( best to create your on 'serialized'  matrix class and and create functions ). 
 
But if You use  ESP32-CAM then  maybe: https://randomnerdtutorials.com/esp32-cam-pir-motion-detector-photo-capture/
will be best solution for you ( + web server app based on openCv python image processing)
